sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/SpannerChangeStreamErrorTest.java

```diff
@@ -0,0 +1,494 @@
+    }
+  }
+
+  @Test
+  public void testAbortedExceptionRetries() {
+    mockSpannerService.setExecuteStreamingSqlExecutionTime(
+        SimulatedExecutionTime.ofStickyException(Status.ABORTED.asRuntimeException()));
+
+    final Timestamp now = Timestamp.now();
+    final Timestamp after3Seconds =
+        Timestamp.ofTimeSecondsAndNanos(now.getSeconds() + 3, now.getNanos());
+    try {
+      pipeline.apply(
+          SpannerIO.readChangeStream()
+              .withSpannerConfig(getSpannerConfig())
+              .withChangeStreamName(TEST_CHANGE_STREAM)
+              .withMetadataDatabase(TEST_DATABASE)
+              .withInclusiveStartAt(now)
+              .withInclusiveEndAt(after3Seconds));
+      pipeline.run().waitUntilFinish();
+    } finally {
+      assertThat(
+          mockSpannerService.countRequestsOfType(ExecuteSqlRequest.class), Matchers.greaterThan(1));
+      thrown.expect(PipelineExecutionException.class);
+      thrown.expectMessage(ErrorCode.ABORTED.name());
+    }
+  }
+
+  @Test
+  public void testUnknownExceptionDoesNotRetry() {
+    mockSpannerService.setExecuteStreamingSqlExecutionTime(
+        SimulatedExecutionTime.ofStickyException(Status.UNKNOWN.asRuntimeException()));
+
+    final Timestamp now = Timestamp.now();
+    final Timestamp after3Seconds =
+        Timestamp.ofTimeSecondsAndNanos(now.getSeconds() + 3, now.getNanos());
+    try {
+      pipeline.apply(
+          SpannerIO.readChangeStream()
+              .withSpannerConfig(getSpannerConfig())
+              .withChangeStreamName(TEST_CHANGE_STREAM)
+              .withMetadataDatabase(TEST_DATABASE)
+              .withInclusiveStartAt(now)
+              .withInclusiveEndAt(after3Seconds));
+      pipeline.run().waitUntilFinish();
+    } finally {
+      assertThat(
+          mockSpannerService.countRequestsOfType(ExecuteSqlRequest.class), Matchers.equalTo(1));
+      thrown.expect(PipelineExecutionException.class);
+      thrown.expectMessage(ErrorCode.UNKNOWN.name());
+    }
+  }
+
+  @Test
+  public void testInvalidRecordReceived() {
+    final Timestamp now = Timestamp.now();
+    final Timestamp after3Seconds =
+        Timestamp.ofTimeSecondsAndNanos(now.getSeconds() + 3, now.getNanos());
+
+    mockTableExists();
+    ResultSet getPartitionResultSet = mockGetParentPartition(now, after3Seconds);
+    mockGetWatermark(now);
+    mockGetPartitionsAfter(
+        Timestamp.ofTimeSecondsAndNanos(now.getSeconds(), now.getNanos() + 1000),
+        getPartitionResultSet);
+    mockGetPartitionsAfter(
+        Timestamp.ofTimeSecondsAndNanos(now.getSeconds(), now.getNanos() - 1000),
+        getPartitionResultSet);
+
+    Statement changeStreamQueryStatement =
+        Statement.newBuilder(
+                "SELECT * FROM READ_my-change-stream(   start_timestamp => @startTimestamp,   end_timestamp => @endTimestamp,   partition_token => @partitionToken,   read_options => null,   heartbeat_milliseconds => @heartbeatMillis)")
```

Comment:
nit: we could remove the extra spacing here (probably spotless added it when merging lines into one)

Reply:
This is actually how it looks in the actual request statement, I believe the test will fail if I take out the spaces — https://github.com/googleapis/java-spanner/blob/main/google-cloud-spanner/src/test/java/com/google/cloud/spanner/MockSpannerServiceImpl.java#L113